pyccel/parser/syntax/openmp.py

```diff
@@ -31,160 +35,686 @@ def __init__(self, **kwargs):
+        _valid_clauses = _valid_taskloop_clauses + (OmpinReduction,)
 
-class OmpLoopConstruct(BasicStmt):
-    """Class representing a ."""
+        txt = self.name
+        txt += check_get_clauses(self, _valid_clauses, self.clauses)
+
+        self._expr = OmpAnnotatedComment(txt)
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> OmpTaskLoopConstruct: expr")
+
+        return self._expr
+
+class OmpTaskConstruct(BasicStmt):
+    """Class representing a Task Construct """
+    def __init__(self, **kwargs):
+        """
+        """
+        self.clauses  = kwargs.pop('clauses')
+        self.name     = kwargs.pop('name')
+
+        _valid_clauses = (OmpPriority,
+                          OmpFinal,
+                          OmpDefault,
+                          OmpPrivate,
+                          OmpShared,
+                          OmpFirstPrivate,
+                          OmpUntied,
+                          OmpMergeable,
+                          OmpinReduction,
+                          OmpDepend)
+
+        txt = self.name
+        txt += check_get_clauses(self, _valid_clauses, self.clauses)
+
+        self._expr = OMP_Task_Construct(txt)
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> OmpTaskConstruct: expr")
+
+        return self._expr
+
+class OmpSingleConstruct(BasicStmt):
+    """Class representing a single construct."""
     def __init__(self, **kwargs):
         """
         """
         self.clauses = kwargs.pop('clauses')
+        self.name    = kwargs.pop('name')
+
+        _valid_clauses = (OmpPrivate,
+                          OmpFirstPrivate)
+
+        txt = self.name
+        txt += check_get_clauses(self, _valid_clauses, self.clauses)
 
-        super(OmpLoopConstruct, self).__init__(**kwargs)
+        self._expr = OMP_Single_Construct(txt)
+
+        super().__init__(**kwargs)
 
     @property
     def expr(self):
         if DEBUG:
-            print("> OmpLoopConstruct: expr")
+            print("> OmpSingleConstruct: expr")
+
+        return self._expr
+
+class OmpCriticalConstruct(BasicStmt):
+    """Class representing a Critical construct."""
+    def __init__(self, **kwargs):
+        """
+        """
+        self.clauses = kwargs.pop('clauses')
+        self.name    = kwargs.pop('name')
+
+        _valid_clauses = (OmpCriticalName)
+
+        txt = self.name
+        txt += check_get_clauses(self, _valid_clauses, self.clauses)
+
+        self._expr = OMP_Critical_Construct(txt)
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> OmpCriticalConstruct: expr")
+
+        return self._expr
+
+class OmpSimdConstruct(BasicStmt):
+    """Class representing a Simd construct."""
+    def __init__(self, **kwargs):
+        """
+        """
+        self.name = kwargs.pop('name')
```

Comment:
Name is not private here, or in other classes. This means that it can be modified outside the class easily. Is this intentional?

Reply:
no it's not intentional, im gonna make it private

Reply:
Ok. If I were you I would fix this last. If you can find a smart solution for the inheritance to avoid code duplication then it should only need doing once instead of in all these different classes